File: Implementations/java/src/test/java/com/google/openlocationcode/EncodingTest.java
package com.google.openlocationcode;

import static java.nio.charset.StandardCharsets.UTF_8;

import java.io.BufferedReader;
import java.io.FileInputStream;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

/** Tests encoding and decoding between Open Location Code and latitude/longitude pair. */
@RunWith(JUnit4.class)
public class EncodingTest {

  public static final double PRECISION = 1e-10;

  private static class TestData {

    private final double latitude;
    private final double longitude;
    private final int length;
    private final String code;

    public TestData(String line) {
      String[] parts = line.split(",");
      if (parts.length != 4) {
        throw new IllegalArgumentException("Wrong format of testing data.");
      }
      this.latitude = Double.parseDouble(parts[0]);
      this.longitude = Double.parseDouble(parts[1]);
      this.length = Integer.parseInt(parts[2]);
      this.code = parts[3];
    }
  }

  private final List<TestData> testDataList = new ArrayList<>();

  @Before
  public void setUp() throws Exception {
    InputStream testDataStream = new FileInputStream(TestUtils.getTestFile("encoding.csv"));
    BufferedReader reader = new BufferedReader(new InputStreamReader(testDataStream, UTF_8));
    String line;
    while ((line = reader.readLine()) != null) {
      if (line.startsWith("#") || line.length() == 0) {
        continue;
      }
      testDataList.add(new TestData(line));
    }
  }

  @Test
  public void testEncodeFromLatLong() {
    for (TestData testData : testDataList) {
      Assert.assertEquals(
          String.format(
              "Latitude %f, longitude %f and length %d were wrongly encoded.",
              testData.latitude, testData.longitude, testData.length),
          testData.code,
          OpenLocationCode.encode(testData.latitude, testData.longitude, testData.length));
    }
  }
}

File: kotlin/src/commonTest/kotlin/com/google/openlocationcode/EncodingTest.kt
package com.google.openlocationcode

import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.fail

class EncodingTest {

    data class TestData(
        val latitude: Double,
        val longitude: Double,
        val length: Int,
        val code: String
    ) {
        companion object {
            fun fromLine(line: String): TestData {
                val parts = line.split(',')
                require(parts.size == 4) { "Wrong format of testing data." }
                return TestData(
                    latitude = parts[0].toDouble(),
                    longitude = parts[1].toDouble(),
                    length = parts[2].toInt(),
                    code = parts[3]
                )
            }
        }
    }

    private lateinit var testDataList: List<TestData>

    @BeforeTest
    fun setUp() {
        testDataList = loadTestCsvFileAsLines("encoding.csv")
            .filterNot { it.startsWith('#') }
            .map { TestData.fromLine(it) }
    }

    @Test
    fun testEncodeFromLatLong() {
        for (testData in testDataList.filter { it.code.isNotBlank() }) {
            assertEquals(
                testData.code,
                OpenLocationCode.encode(testData.latitude, testData.longitude, testData.length),
                "Latitude ${testData.latitude}, longitude ${testData.longitude} and length ${testData.length} were wrongly encoded."
            )
        }
    }

    @Test
    fun testEncodeThrowsIllegalArgumentException() {
        try {
            for (testData in testDataList.filter { it.code.isBlank() }) {
                OpenLocationCode.encode(testData.latitude, testData.longitude, testData.length);
            }
            fail("Expected IllegalArgumentException")
        } catch (ex: IllegalArgumentException) {
            // pass!
        }
    }

}
